### PR TITLE
Even better Makefile target parsing

### PR DESCRIPTION
**Pull request by adamjstewart**
This is a continuation of #8223.

In #8818, @SteVwonder discovered a bug in my implementation. For reasons I still don't fully understand, `make -q` does not work as documented. This can easily be reproduced with the `libsigsegv` package, which does not have any dependencies. Outside of Spack, we can see that `make check` works fine, but `make -q check` crashes:
```console
$ make -q check
Making check in src
Making check in tests
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
make[1]: *** [check-am] Error 1
make: *** [check-recursive] Error 1
$ echo $?
2
$ make check
Making check in src
make[1]: Nothing to be done for `check'.
Making check in tests
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
PASS: sigsegv1
PASS: sigsegv2
PASS: sigsegv3
PASS: stackoverflow1
PASS: stackoverflow2
============================================================================
Testsuite summary for 
============================================================================
# TOTAL: 5
# PASS:  5
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================

Now please type 'make install' to install the package.
$ echo $?
0
```
The error code for `make -q` was also unreliable, it only worked for GNU Make, not BSD Make.

The new implementation ignores the exit code and instead parses STDERR for any indications that the target does not exist. This works for both GNU Make and BSD Make.

I switched from `make -q` to `make -n`. It works similarly to `make -q` (it doesn't actually run anything, just prints what would be run). The only difference is that `make -q` doesn't print an error message when using BSD Make, but `make -n` does.

@SteVwonder would you mind testing this with #8818?

Closes #8604

### CODE REVIEW

**Comment by SteVwonder** [comment]
> For reasons I still don't fully understand, make -q does not work as documented.

I think `-q` is working as documented.
```
      -q, --question
            ``Question mode''.  Do not run any commands, or print anything; just return an exit status that is zero if the specified targets are already up to date, nonzero otherwise.
```
It only returns an exit status of zero if the specified target is up to date (i.e., no commands need to be run).  This will never be the case with `make check` because `check` isn't a "real" target, it is a "phony" one that is designed to always execute commands.  So it is always out of date.  I don't think the `--question` flag is broken, I think it is just asking a different question than we want.

Concrete examples, in both cases, the target exists.  In only the last case is the target up to date.
```
# herbein1 at hype2 in ~/Repositories/flux-framework/flux-sched/src/common on git:upmaster ✖︎ [20:51:25]
→ make -q; echo "$?"
Making all in libtap
Making all in liblsd
make[1]: Entering directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/liblsd'
make[1]: Warning: File `.deps/list.Plo' has modification time 6.2 s in the future
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/liblsd'
make: *** [all-recursive] Error 1
2

# herbein1 at hype2 in ~/Repositories/flux-framework/flux-sched/src/common on git:upmaster ✖︎ [20:51:29]
→ make; echo "$?"
Making all in libtap
make[1]: Entering directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/libtap'
make[1]: Nothing to be done for `all'.
make[1]: Leaving directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/libtap'
Making all in liblsd
make[1]: Entering directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/liblsd'
make[1]: Warning: File `.deps/list.Plo' has modification time 2 s in the future
  CC       list.lo
  CCLD     liblsd.la
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/liblsd'
Making all in libutil
make[1]: Entering directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/libutil'
make[1]: Warning: File `.deps/xzmalloc.Plo' has modification time 1.8 s in the future
  CC       xzmalloc.lo
  CC       log.lo
  CCLD     libutil.la
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/libutil'
Making all in librbtree
make[1]: Entering directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/librbtree'
make[1]: Warning: File `.deps/rbtree.Plo' has modification time 2 s in the future
  CC       rbtree.lo
  CCLD     librbtree.la
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/librbtree'
make[1]: Entering directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common'
make[1]: Nothing to be done for `all-am'.
make[1]: Leaving directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common'
0

# herbein1 at hype2 in ~/Repositories/flux-framework/flux-sched/src/common on git:upmaster ✖︎ [20:51:34]
→ make -q; echo "$?"
Making all in libtap
Making all in liblsd
make[1]: Entering directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/liblsd'
make[1]: Warning: File `.deps/list.Plo' has modification time 12 s in the future
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/liblsd'
Making all in libutil
make[1]: Entering directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/libutil'
make[1]: Warning: File `.deps/xzmalloc.Plo' has modification time 12 s in the future
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/libutil'
Making all in librbtree
make[1]: Entering directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/librbtree'
make[1]: Warning: File `.deps/rbtree.Plo' has modification time 13 s in the future
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/g/g0/herbein1/Repositories/flux-framework/flux-sched/src/common/librbtree'
0
```

**Comment by SteVwonder** [comment]
> @SteVwonder would you mind testing this with #8818?

I gave it a whirl.  This is the output I could see:
```
==> RUN-TESTS: build-time tests [check]
==> 'make' '-j56' '-n' 'test'
==> Target 'test' not found in Makefile
==> 'make' '-j56' '-n' 'check'
==> Target 'check' not found in Makefile
==> Executing phase: 'install'
==> 'make' '-j56' 'install'
```

I'm digging deeper now.

**Comment by SteVwonder** [comment]
Output from manually running it:
```
# herbein1 at hype2 in /usr/workspace/wsb/herbein1/spack_toss3/var/spack/stage/flux-core-master-ax7lswmzn2koloklj5zinph2efc2acqr/flux-core/spack-build on git:master ✖︎ [21:04:48]
→ make -n check
fail=; \
if (target_option=k; case ${target_option-} in ?) ;; *) echo "am__make_running_with_option: internal error: invalid" "target option '${target_option-}' specified" >&2; exit 1;; esac; has_opt=no; sane_makeflags=$MAKEFLAGS; if { if test -z '
0'; then false; elif test -n ''; then true; elif test -n '3.82' && test -n '/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build'; then true; else false; fi; }; then sane_makeflags=$MFLAGS; else case $MAKEFLAGS in *\\[\ \   ]
*) bs=\\; sane_makeflags=`printf '%s\n' "$MAKEFLAGS" | sed "s/$bs$bs[$bs $bs    ]*//g"`;; esac; fi; skip_next=no; strip_trailopt () { flg=`printf '%s\n' "$flg" | sed "s/$1.*$//"`; }; for flg in $sane_makeflags; do test $skip_next = yes &&
{ skip_next=no; continue; }; case $flg in *=*|--*) continue;; -*I) strip_trailopt 'I'; skip_next=yes;; -*I?*) strip_trailopt 'I';; -*O) strip_trailopt 'O'; skip_next=yes;; -*O?*) strip_trailopt 'O';; -*l) strip_trailopt 'l'; skip_next=yes;
; -*l?*) strip_trailopt 'l';; -[dEDm]) skip_next=yes;; -[JT]) skip_next=yes;; esac; case $flg in *$target_option*) has_opt=yes; break;; esac; done; test $has_opt = yes); then \
  failcom='fail=yes'; \
else \
  failcom='exit 1'; \
fi; \
dot_seen=no; \
target=`echo check-recursive | sed s/-recursive//`; \
case "check-recursive" in \
  distclean-* | maintainer-clean-*) list='. src doc etc t' ;; \
  *) list='. src doc etc t' ;; \
esac; \
for subdir in $list; do \
  echo "Making $target in $subdir"; \
  if test "$subdir" = "."; then \
    dot_seen=yes; \
    local_target="$target-am"; \
  else \
    local_target="$target"; \
  fi; \
  (CDPATH="${ZSH_VERSION+.}:" && cd $subdir && make  $local_target) \
  || eval $failcom; \
done; \
if test "$dot_seen" = "no"; then \
  make  "$target-am" || exit 1; \
fi; test -z "$fail"
Making check in .
make[1]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build'
make  check-local
make[2]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build'

<snip>

make[4]: Leaving directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build/src/common/libtap'
make[3]: Leaving directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build/src/common/libtap'
Making check in libev
make[3]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build/src/common/libev'
make[3]: Nothing to be done for `check'.
make[3]: Leaving directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build/src/common/libev'
Making check in libminilzo
make[3]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build/src/common/libminilzo'
make  test_mini.t
make[4]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build/src/common/libminilzo'
/usr/bin/mkdir -p test
: > test/.dirstamp
/usr/bin/mkdir -p test/.deps
: > test/.deps/.dirstamp
echo "  CC      " test/mini.o;depbase=`echo test/mini.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
/usr/workspace/wsb/herbein1/spack_toss3/lib/spack/env/gcc/gcc -std=gnu99 -DHAVE_CONFIG_H -I. -I/usr/workspace/wsb/herbein1/spack_toss3/var/spack/stage/flux-core-master-ax7lswmzn2koloklj5zinph2efc2acqr/flux-core/src/common/libminilzo -I../.
./../config  -I/usr/workspace/wsb/herbein1/spack_toss3/var/spack/stage/flux-core-master-ax7lswmzn2koloklj5zinph2efc2acqr/flux-core -I/usr/workspace/wsb/herbein1/spack_toss3/var/spack/stage/flux-core-master-ax7lswmzn2koloklj5zinph2efc2acqr/
flux-core/src/include -Wno-strict-aliasing -Wno-error=strict-aliasing  -Wall -Werror -Werror=missing-field-initializers -Wno-strict-aliasing -Wno-error=deprecated-declarations  -g -O2 -MT test/mini.o -MD -MP -MF $depbase.Tpo -c -o test/min
i.o /usr/workspace/wsb/herbein1/spack_toss3/var/spack/stage/flux-core-master-ax7lswmzn2koloklj5zinph2efc2acqr/flux-core/src/common/libminilzo/test/mini.c &&\
mv -f $depbase.Tpo $depbase.Po
make[4]: *** No rule to make target `../../../src/common/libtap/libtap.la', needed by `test_mini.t'.  Stop.
make[4]: Leaving directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build/src/common/libminilzo'
make[3]: *** [check-am] Error 2
make[3]: Leaving directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build/src/common/libminilzo'
make[2]: *** [check-recursive] Error 1
make[2]: Leaving directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build/src/common'
make[1]: *** [check-recursive] Error 1
make[1]: Leaving directory `/tmp/herbein1/spack-stage/spack-stage-lhoEHx/flux-core/spack-build/src'
make: *** [check-recursive] Error 1
```

**Comment by SteVwonder** [comment]
Just a heads up, it appears that `-n` might still execute some receipes.  Per the [online manual](https://www.gnu.org/software/make/manual/html_node/Instead-of-Execution.html#Instead-of-Execution):
```
‘-n’
‘--just-print’
‘--dry-run’
‘--recon’

    “No-op”. Causes make to print the recipes that are needed to make the targets up to date, but not actually execute them. Note that some recipes are still executed, even with this flag (see How the MAKE Variable Works). Also any recipes needed to update included makefiles are still executed (see How Makefiles Are Remade).
```

**Comment by SteVwonder** [comment]
It seems that most of the trouble with flux-core/sched's Makefile's comes from the subdirectories and the recursive nature of the calls.  **Potentially stupid idea alert**: what if we just use grep on the top-level Makefile?
```
 herbein1 at hype2 in /usr/workspace/wsb/herbein1/spack_toss3/var/spack/stage/flux-core-master-ax7lswmzn2koloklj5zinph2efc2acqr/flux-core/spack-build on git:master ✖︎ [21:10:47]
→ grep -q "^check:" ./Makefile && echo "check target exists" || echo "NO CHECK TARGET"
check target exists
```

**Comment by adamjstewart** [comment]
See #8223 for why grep wasn't sufficient. If we can come up with a better regex, that may be the last resort.

**Comment by adamjstewart** [comment]
@SteVwonder Can you try the latest commit? That should fix the problem you were seeing. I still don't like the fact that it errors out though.

We need to decide how worrying the potential for `make -n` to run the Makefile is. If this is a big concern, we will need to go back to regexes. For example, if we are looking for the `check` target, we would need to be able to capture:
```make
check:
check :
test check:
check test:
foo bar check baz:
```
but not:
```make
installcheck:
checkinstall:
```

**Comment by SteVwonder** [comment]
@adamjstewart: your latest commit seems to work:

```
==> RUN-TESTS: build-time tests [check]
==> 'make' '-j56' '-n' 'test'
==> Target 'test' not found in Makefile
==> 'make' '-j56' '-n' 'check'
==> 'make' '-j56' 'check'
Making check in .
make[1]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-Blgqv1/flux-core/spack-build'
make  check-local
make[2]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-Blgqv1/flux-core/spack-build'
Making all in .
make[3]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-Blgqv1/flux-core/spack-build'
make[3]: Nothing to be done for `all-am'.
make[3]: Leaving directory `/tmp/herbein1/spack-stage/spack-stage-Blgqv1/flux-core/spack-build'
Making all in src
make[3]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-Blgqv1/flux-core/spack-build/src'
Making all in common
make[4]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-Blgqv1/flux-core/spack-build/src/common'
Making all in libtap
make[5]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-Blgqv1/flux-core/spack-build/src/common/libtap'
make[5]: Nothing to be done for `all'.
make[5]: Leaving directory `/tmp/herbein1/spack-stage/spack-stage-Blgqv1/flux-core/spack-build/src/common/libtap'
Making all in libev
make[5]: Entering directory `/tmp/herbein1/spack-stage/spack-stage-Blgqv1/flux-core/spack-build/src/common/libev'
make[5]: Nothing to be done for `all'.
make[5]: Leaving directory `/tmp/herbein1/spack-stage/spack-stage-Blgqv1/flux-core/spack-build/src/common/libev'

<snip>
```
FWIW, `^([a-z]+ )*check( [a-z]+)* *:$` should work for the test cases you provided: https://regex101.com/r/e3LuE4/1

Another potentially terrible idea is to actually parse the Makefile using pyparsing or similar.

**Comment by adamjstewart** [comment]
I might try that regex. I need to find some documentation on allowed target names (I think they can contain uppercase and dashes) but that might be just what I'm looking for. Avoiding accidentally running the Makefile or dealing with different versions of Make sounds nice.

**Comment by SteVwonder** [comment]
It looks like targets can contain:
- (any?) valid filenames
- wildcards
- parantheses (e.g., `a(m)`)
- percent sign
- (the tricky one): variables (e.g., `$(OBJS):`)
- other stuff?
source: https://www.gnu.org/software/make/manual/make.html#Rule-Syntax

We could handle a lot of that via the regex character class `[^:]`, but I'm not sure how we would handle the case of variables expanding to include `check`:
```
# herbein1 at auk100.llnl.gov in ~/Repositories/ on git:master ✖︎ [13:37:05]
→ cat Makefile  
TARGETS=check

$(TARGETS):
        echo "RUNNING"

# herbein1 at auk100.llnl.gov in ~/Repositories/ on git:master ✖︎ [13:37:14]
→ make check
echo "RUNNING"
RUNNING
```

The `make -n` case does handle the variable expansion though:
```
# herbein1 at auk100.llnl.gov in ~/Repositories/ on git:master ✖︎ [13:37:25]
→ make -n check
echo "RUNNING"

# herbein1 at auk100.llnl.gov in ~/Repositories/ on git:master ✖︎ [13:41:24]
→ make -n foobar
make: *** No rule to make target `foobar'.  Stop.
```

**Comment by adamjstewart** [comment]
@SteVwonder So far your regex is up to:
```
^([^ :]+ )*check( [^ :]+)* *:$
```
There are two problems with this:

1. The line doesn't have to end with a colon. We could have a file like:

```make
check: recursive-check

recursive-check:
        the actual build instructions
```
So we need to drop the `$`.

2. We need to avoid some other things, like variable assignment:

```make
check := ${test}
```

Does this seem sufficient?
```
^([^ :]+ )*check( [^ :]+)* *:[^=]
```
I wouldn't think this would match a line ending with a colon, but regex101.com seems to think otherwise.

P.S. This is why I gave up last time. So many corner cases...

**Comment by SteVwonder** [comment]
@adamjstewart: I'm leaning towards sticking with `make -n`.  The regex is going to be fragile compared to relying on `make` to spit out the target names.  It looks like the main case where `make -n` still executes recipes is when recursive calls to `make` are made without using the `${MAKE}` variable.  Spack only calls `make -n` after having already run `make`, correct?  So I **assume** that recipes will just be running for the second time if things actually do get run (resulting in a no-op).  The caveat being that I am not sure if there are additional scenarios where `make -n` causes things to actually execute.    And I'm sure there are some really twisted Makefiles out there that violate all these assumptions (in those cases, I guess we just override the `check` method).

If you want to go with the regex, I can take another look at what you have so far.

**Comment by adamjstewart** [comment]
Ping

File: lib/spack/spack/test/data/ninja/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+.ninja_deps
+.ninja_log
```
Comment (adamjstewart) [inline lib/spack/spack/test/data/ninja/.gitignore]: Running `ninja check` generates these files. Should I try to clean them up? How do I do that in pytest?

**Comment by adamjstewart** [comment]
Okay, this should be ready to merge now. I added unit tests for every conceivable `Makefile` or `build.ninja` we might encounter. It provides a nice boost to our core coverage as well.

Let me know if you want me to clean up those `.ninja_*` files.

**Comment by scheibelp** [comment]
Are you still tweaking this? It looked good to me so I was going to merge it

**Comment by adamjstewart** [comment]
I'm done, I was just trying to clean up those pesky Ninja files. If the last commit looks good to you go ahead and merge.

**Comment by scheibelp** [comment]
It looks good. However, I'm writing a commit message and I'm curious if you added https://github.com/spack/spack/pull/8819/commits/0222d1e8d50ac8b27e62b1b524c5b72aaf4dc445 (adding 'all' to the `ninja` invocation) only to be more explicit - i.e. is the behavior the same with or without 'all'? could it conceivably fail or return something unexpected without 'all'?

**Comment by adamjstewart** [comment]
I added it only to be more explicit. The behavior is identical with or without `all`. For reference, [Ninja's Bash completion script](https://github.com/ninja-build/ninja/blob/master/misc/bash-completion#L52) also uses `ninja -t targets all`.

**Comment by scheibelp** [comment]
It appears that a couple of the `test_negative_make_check` tests failed in https://travis-ci.org/spack/spack/jobs/418414627 (for python 3.7) - do you have a notion of why this would have occurred?

**Comment by adamjstewart** [comment]
No idea.

**Comment by scheibelp** [comment]
OK - I have added an issue on 3.7 test failures at https://github.com/spack/spack/issues/9028. For now, we allow those to fail so I'll merge this.

**Comment by tgamblin** [comment]
@adamjstewart @scheibelp: is `develop` failing on this PR?

**Comment by alalazo** [comment]
> Develop seems to be passing? It only fails the Python 3.7 test.

Imo, any test failing on Python 3.7 should be considered a "real" failure. In #8778 I added:
```yaml
allow_failures:
    - dist: xenial
```
just because the way to get a worker on Travis with `xenial` is not yet official, and it may change without notice in the future (for the record using `xenial` seemed to be the only easy way to run on Python 3.7). See also https://github.com/travis-ci/travis-ci/issues/9815

**Pull request merged**